test/e2e/daemon_set.go

```diff
@@ -59,37 +64,74 @@ var _ = Describe("Daemon set", func() {
 	})
 })
 
-func clearNodeLabels(c *client.Client) error {
+func separateDaemonSetNodeLabels(labels map[string]string) (map[string]string, map[string]string) {
+	daemonSetLabels := map[string]string{}
+	otherLabels := map[string]string{}
+	for k, v := range labels {
+		if strings.HasPrefix(k, daemonsetLabelPrefix) {
+			daemonSetLabels[k] = v
+		} else {
+			otherLabels[k] = v
+		}
+	}
+	return daemonSetLabels, otherLabels
+}
+
+func clearDaemonSetNodeLabels(c *client.Client) error {
 	nodeClient := c.Nodes()
 	nodeList, err := nodeClient.List(labels.Everything(), fields.Everything())
 	if err != nil {
 		return err
 	}
 	for _, node := range nodeList.Items {
-		if len(node.Labels) != 0 {
-			node.Labels = map[string]string{}
-			var newNode *api.Node
-			err = wait.Poll(updateRetryPeriod, updateRetryTimeout, func() (bool, error) {
-				newNode, err = nodeClient.Update(&node)
-				if err == nil {
-					return true, err
-				}
-				if se, ok := err.(*apierrs.StatusError); ok && se.ErrStatus.Reason == unversioned.StatusReasonConflict {
-					Logf("failed to update node due to resource version conflict")
-					return false, nil
-				}
-				return false, err
-			})
-			if err != nil {
-				return err
-			} else if len(newNode.Labels) != 0 {
-				return fmt.Errorf("Could not make node labels nil.")
-			}
+		_, err := setDaemonSetNodeLabels(c, node.Name, map[string]string{})
+		if err != nil {
+			return err
 		}
 	}
 	return nil
 }
 
+func setDaemonSetNodeLabels(c *client.Client, nodeName string, labels map[string]string) (*api.Node, error) {
+	nodeClient := c.Nodes()
+	var newNode *api.Node
+	var newLabels map[string]string
+	err := wait.Poll(updateRetryPeriod, updateRetryTimeout, func() (bool, error) {
+		node, err := nodeClient.Get(nodeName)
+		if err != nil {
+			return false, err
+		}
+
+		// remove all labels this test is creating
+		daemonSetLabels, otherLabels := separateDaemonSetNodeLabels(node.Labels)
+		if reflect.DeepEqual(daemonSetLabels, labels) {
+			newNode = node
+			return true, nil
+		}
+		node.Labels = otherLabels
+		for k, v := range labels {
+			node.Labels[k] = v
+		}
+		newNode, err = nodeClient.Update(node)
+		if err == nil {
+			newLabels, _ = separateDaemonSetNodeLabels(newNode.Labels)
+			return true, err
+		}
+		if se, ok := err.(*apierrs.StatusError); ok && se.ErrStatus.Reason == unversioned.StatusReasonConflict {
+			Logf("failed to update node due to resource version conflict")
+			return false, nil
+		}
+		return false, err
+	})
+	if err != nil {
+		return nil, err
+	} else if len(newLabels) != len(labels) {
+		return nil, fmt.Errorf("Could not remove daemon set test node labels.")
```

Comment:
nit: error message seems incorrect. Is this method used only for removing labels?